### PR TITLE
Verify JDBC driver property precedence and remove default values

### LINKED ISSUE #26206: SQLServer driver property precedence
## Current Behavior

If I configure the following DataSource for SQLServer: 

```xml
<dataSource jndiName="jdbc/with-url">
    <jdbcDriver libraryRef="sqlserverlib"/>
    <properties.microsoft.sqlserver user="${env.SQL_USER}" password="${env.SQL_PASSWORD}" 
    	url="jdbc:sqlserver://my.host.com:1433;databaseName=TEST;" />
</dataSource>
```

I would expect the SQLServer Driver to create a connection to my database using the serverName `my.host.com`.

However, the `properties.microsoft.sqlserver` has a default value configured for the serverName `localhost`:

https://github.com/OpenLiberty/open-liberty/blob/969616da96ff8700d3d06fd7b77b9a361925d7e3/dev/com.ibm.ws.jdbc.metatype/resources/OSGI-INF/metatype/metatype.xml#L1198

As a result, the SQLServer Driver attempts to create a connection to a database using the serverName `localhost` and fails.

This is because the SQLServer Driver takes supplied properties as precedence of URL properties:

https://github.com/microsoft/mssql-jdbc/blob/145c10e90148087e48c6c18f0f3449bcbb3d1708/src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java#L1110-L1138

## Work around

Therefore, the only way for me to use a URL is configure both the URL and serverName property: 

```xml
<dataSource jndiName="jdbc/with-url">
    <jdbcDriver libraryRef="sqlserverlib"/>
    <properties.microsoft.sqlserver user="${env.SQL_USER}" password="${env.SQL_PASSWORD}" 
    	url="jdbc:sqlserver://my.host.com:1433;databaseName=TEST;" serverName="my.host.com"/>
</dataSource>
```

## Proposed behavior

~~I would propose that if a URL and default serverName from metatype are configured, then we should ignore the default serverName from metatype and only set the URL on the JDBC driver.~~

~~If a URL and serverName are configured, the we should pass each to the JDBC driver and let the JDBC driver choose which to use based on it's precedence.~~

- We remove the default value `localhost` from `serverName` in metatype.
- Update metatype to indicate
    - If `serverName` `portNumber` `instanceName` or `databaseName` attributes are configured
    - then they will overwrite the corresponding values if they were also configured in the url.
- To minimize this breaking change we would add the following logic when creating a DataSource or Driver: 
  - If `url` is not configured, `serverName` is not configured, and `properties.microsoft.sqlserver` is used
  - then set `serverName` to `localhost` and pass it to the JDBC driver.

## Breaking change
Who would this change break? 

If someone configured their server with only a URL:
```xml
<dataSource jndiName="jdbc/with-url">
    <jdbcDriver libraryRef="sqlserverlib"/>
    <properties.microsoft.sqlserver user="${env.SQL_USER}" password="${env.SQL_PASSWORD}" 
    	url="jdbc:sqlserver://my.host.com:1433;databaseName=TEST;" />
</dataSource>
```

And that the hostname was switched to `localhost`: `jdbc:sqlserver://localhost:1433;databaseName=TEST;`
And there happened to be an instance of SQLServer available at that address.
Then this change would break this configuration because the user would start connecting to the configured host `my.host.com`

## Design Issue Meeting Minutes

1. We should document the precedence of properties for all drivers when there is a ambiguous decision made by the JDBC Driver.
2. We refined the proposed solution for better clarity and minimization of breaking changes.
3. Bringing this issue to a wider audience to have a vote for approval.

### PR DESCRIPTION
**Pull request by KyleAure**
Fixes #26206 

Verifies the behavior of JDBC drivers when both a URL and connection properties (serverName, portNumber, databaseName) are set. 

In the case of SQLServer and DB2 we want to avoid setting default values, because those JDBC drivers will override the URL with those default values.

### CODE REVIEW

File: dev/com.ibm.ws.jdbc.metatype/resources/OSGI-INF/l10n/metatype.properties [hunk 1]
```diff
@@ -371,6 +371,9 @@ cursorSensitivity=Cursor sensitivity
 cursorSensitivity.desc=JDBC driver property: cursorSensitivity. Values are: 0 (TYPE_SCROLL_SENSITIVE_STATIC), 1 (TYPE_SCROLL_SENSITIVE_DYNAMIC), 2 (TYPE_SCROLL_ASENSITIVE).
 databaseName=Database name
 databaseName.desc=JDBC driver property: databaseName.
+databaseNameIgnored.desc=JDBC driver property: databaseName. This property is ignored by the driver if a URL is configured.
+databaseNameOverrides.desc=JDBC driver property: databaseName. This property will override the databaseName specified in the URL if one was configured.
```
Comment (chirp1) [inline dev/com.ibm.ws.jdbc.metatype/resources/OSGI-INF/l10n/metatype.properties]: Change "This property will override the databaseName specified..." to "This property overrides the databaseName property specified...".

File: dev/com.ibm.ws.jdbc.metatype/resources/OSGI-INF/l10n/metatype.properties [hunk 2]
```diff
@@ -659,6 +662,7 @@ insensitiveResultSetBufferSize=Insensitive result set buffer size
 insensitiveResultSetBufferSize.desc=JDBC driver property: insensitiveResultSetBufferSize.
 instanceName=Instance name
 instanceName.desc=JDBC driver property: instanceName.
+instanceNameOverrides.desc=JDBC driver property: instanceName. This property will override the instanceName specified in the URL if one was configured.
```
Comment (chirp1) [inline dev/com.ibm.ws.jdbc.metatype/resources/OSGI-INF/l10n/metatype.properties]: Change "This property will override..." to "This property overrides...".

File: dev/com.ibm.ws.jdbc.metatype/resources/OSGI-INF/l10n/metatype.properties [hunk 3]
```diff
@@ -781,6 +785,9 @@ pkList=Package list
 pkList.desc=JDBC driver property: pkList.
 portNumber=Port number
 portNumber.desc=Port on which to obtain database connections.
+portNumberIgnored.desc=JDBC driver property: portNumber. This property is ignored by the driver if a URL is configured.
+portNumberOverride.desc=JDBC driver property: portNumber. This property will override the portNumber specified in the URL if one was configured.
```
Comment (chirp1) [inline dev/com.ibm.ws.jdbc.metatype/resources/OSGI-INF/l10n/metatype.properties]: Change to "This property overrides the portNumber property specified....".

File: dev/com.ibm.ws.jdbc.metatype/resources/OSGI-INF/l10n/metatype.properties [hunk 4]
```diff
@@ -885,6 +892,9 @@ serverInitTransactions=Server initiated transactions
 serverInitTransactions.desc=JDBC driver property: SERVER_INITIATED_TRANSACTIONS.
 serverName=Server name
 serverName.desc=Server where the database is running.
+serverNameIgnored.desc=JDBC driver property: serverName. This property is ignored by the driver if a URL is configured.
+serverNameOverrides.desc=JDBC driver property: serverName. This property will override the serverName specified in the URL if one was configured.
```
Comment (chirp1) [inline dev/com.ibm.ws.jdbc.metatype/resources/OSGI-INF/l10n/metatype.properties]: Change to "This property overrides the serverName property specified in...".

File: dev/com.ibm.ws.jdbc.metatype/resources/OSGI-INF/l10n/metatype.properties [hunk 5]
```diff
@@ -885,6 +892,9 @@ serverInitTransactions=Server initiated transactions
...
+serverNamePostgreSQL.desc=JDBC driver property: serverName. If a URL and this property are configured, then the PostgreSQL driver will ignore this property when using type java.sql.Driver, and ignore the serverName from the URL when using type javax.sql.DataSource.
```
Comment (chirp1) [inline dev/com.ibm.ws.jdbc.metatype/resources/OSGI-INF/l10n/metatype.properties]: Change to:
When the PostgreSQL driver uses type java.sql.Driver and a URL and this property are configured, the PostgreSQL driver ignores this property. When the PostgreSQL driver uses type javax.sql.DataSource, and a URL and this property are configured,  the PostgreSQL driver ignores the serverName property.

File: dev/com.ibm.ws.jdbc.metatype/resources/OSGI-INF/l10n/metatype.properties [hunk 6]
```diff
@@ -781,6 +785,9 @@ pkList=Package list
 pkList.desc=JDBC driver property: pkList.
 portNumber=Port number
 portNumber.desc=Port on which to obtain database connections.
+portNumberIgnored.desc=JDBC driver property: portNumber. This property is ignored by the driver if a URL is configured.
+portNumberOverride.desc=JDBC driver property: portNumber. This property will override the portNumber specified in the URL if one was configured.
+portNumberPostgreSQL.desc=JDBC driver property: portNumber. If a URL and this property are configured, then the PostgreSQL driver will ignore this property when using type java.sql.Driver, and ignore the portNumber from the URL when using type javax.sql.DataSource.
```
Comment (chirp1) [inline dev/com.ibm.ws.jdbc.metatype/resources/OSGI-INF/l10n/metatype.properties]: Change to:
When the PostgreSQL driver uses type java.sql.Driver and a URL and this property are configured, the PostgreSQL driver ignores this property. When the PostgreSQL driver uses type javax.sql.DataSource, and a URL and this property are configured, the PostgreSQL driver ignores the portNumber property.

File: dev/com.ibm.ws.jdbc.metatype/resources/OSGI-INF/l10n/metatype.properties [hunk 7]
```diff
@@ -371,6 +371,9 @@ cursorSensitivity=Cursor sensitivity
 cursorSensitivity.desc=JDBC driver property: cursorSensitivity. Values are: 0 (TYPE_SCROLL_SENSITIVE_STATIC), 1 (TYPE_SCROLL_SENSITIVE_DYNAMIC), 2 (TYPE_SCROLL_ASENSITIVE).
 databaseName=Database name
 databaseName.desc=JDBC driver property: databaseName.
+databaseNameIgnored.desc=JDBC driver property: databaseName. This property is ignored by the driver if a URL is configured.
+databaseNameOverrides.desc=JDBC driver property: databaseName. This property will override the databaseName specified in the URL if one was configured.
+databaseNamePostgreSQL.desc=JDBC driver property: databaseName. If a URL and this property are configured, then the PostgreSQL driver will ignore this property when using type java.sql.Driver, and ignore the databaseName from the URL when using type javax.sql.DataSource.
```
Comment (chirp1) [inline dev/com.ibm.ws.jdbc.metatype/resources/OSGI-INF/l10n/metatype.properties]: Change to:
When the PostgreSQL driver uses type java.sql.Driver and a URL and this property are configured, the PostgreSQL driver ignores this property. When the PostgreSQL driver uses type javax.sql.DataSource, and a URL and this property are configured, the PostgreSQL driver ignores the databaseName property.

**Review by chirp1 — COMMENTED** [review]
Hi! I provided comments for the messages. If you have any questions or want to discuss my review comments, let me know.

**Comment by posmith** [comment]
L2 message review not required.

**Comment by KyleAure** [comment]
#libby recheck headers

**Comment by KyleAure** [comment]
@chirp1 -- Updates have been made based on feedback.

**Review by njr-11 — DISMISSED** [review]

**Comment by KyleAure** [comment]
#build

**Comment by chirp1** [comment]
@KyleAure Updates look good. I added the "ID reviewed" label.

**Comment by KyleAure** [comment]
#build - redo build

**Comment by KyleAure** [comment]
#libby - recheck headers
#build - redo personal build

**Comment by KyleAure** [comment]
Personal build passed, but was never reported back by Libby, see results: 
https://wasrtc.hursley.ibm.com:9443/jazz/resource/itemOid/com.ibm.team.build.BuildResult/_UvzaQG6REe6Qs_3uoOzbQw

Target locations of links might be accessible only to IBM employees.

**Review by njr-11 — APPROVED** [review]

**Pull request merged**